core/silkworm/consensus/engine.hpp

```diff
@@ -0,0 +1,94 @@
+    virtual ~IConsensusEngine() = default;
+
+    //! \brief Performs validation of block header & body that can be done prior to sender recovery and execution.
+    //! \brief See [YP] Sections 4.3.2 "Holistic Validity", 4.3.4 "Block Header Validity", and 11.1 "Ommer Validation".
+    //! \param [in] block: block to pre-validate.
+    //! \param [in] state: current state.
+    //! \note Shouldn't be used for genesis block.
+    virtual ValidationResult validate_block(const Block& block, State& state) = 0;
+
+    //! \brief See [YP] Section 4.3.4 "Block Header Validity".
+    //! \param [in] header: header to validate.
+    //! \param [in] state: current state.
+    //! \note Shouldn't be used for genesis block.
+    virtual ValidationResult validate_block_header(const BlockHeader& header, State& state) = 0;
+
+    //! \brief Validates the seal of the header
+    virtual ValidationResult validate_seal(const BlockHeader& header) {
+        (void)header;
+        return ValidationResult::kOk;
+    }
+
+    //! \brief Finalizes consensus tasks by applying changes in the state of accounts or of the consensus itself
+    //! \param [in] state: current state.
+    //! \param [in] block: current block to apply rewards for.
+    //! \param [in] revision: EVM fork.
+    //! \remarks For Ethash See [YP] Section 11.3 "Reward Application".
+    virtual void finalize(IntraBlockState& state, const Block& block, const evmc_revision& revision);
+
+    //! \brief See [YP] Section 11.3 "Reward Application".
+    //! \param [in] header: Current block to get beneficiary from
+    virtual evmc::address get_beneficiary(const BlockHeader& header) = 0;
+
+    //! \brief Returns parent header (if any) of provided header
+    virtual std::optional<BlockHeader> get_parent_header(const State& state, const BlockHeader& header) = 0;
+
+  protected:
+    const ChainConfig& chain_config_;
+
+    //! \brief See [YP] Section 11.1 "Ommer Validation"
+    virtual bool is_kin(const BlockHeader& branch_header, const BlockHeader& mainline_header,
+                        const evmc::bytes32& mainline_hash, unsigned n, const State& state,
+                        std::vector<BlockHeader>& old_ommers) = 0;
+
+    //! \brief See https://eips.ethereum.org/EIPS/eip-1559
+    virtual std::optional<intx::uint256> expected_base_fee_per_gas(const BlockHeader& header,
+                                                                   const BlockHeader& parent) = 0;
+};
+
+//! \brief Performs a validation of a transaction.
+//! \return Any of kIntrinsicGas, kInvalidSignature, kWrongChainId, kUnsupportedTransactionType, or kOk.
+//! \remarks Should sender of transaction not yet recovered a check on signature's validity is performed
+//! \remarks These function is agnostic to whole block validity
+ValidationResult validate_transaction(const Transaction& txn, uint64_t block_number, const ChainConfig& config,
```

Comment:
I'd keep `pre_validate_transaction` and the comment "Performs validation of a transaction that can be done prior to sender recovery and block execution" because it's not the full validation. There are also checks in `ExecutionProcessor::validate_transaction`.

Reply:
I understand the point but let me argue that :

- in the context of "consensus_engine" this is the only validation: any "pre" (or "post") suffix implies there is also a "non-pre" (or "non-post") validation while instead there are none in **the context**
- I understand the concept of consensus in the chain involves both the "grammar and syntax" of blocks and transactions (which actually is what consensus_engine checks) **bundled** with the correctness of the state which is delegated to execution phase(s)

This said we're in a situation where naming is tricky. 
Wrt `validate_transaction` actual code in `processor.cpp` only invokes it as assertion (which is skipped in release build) so, again, the "pre" prefix does not make much sense.

I am also assuming that in a "stage-loop" context like ours "consensus_engine" checks come before execution.

So either we rename `consensus_engine` to `pre_execution_consensus_engine` or we restore the "pre" suffix (which I definitely don't like but eventually can live with).

Reply:
`pre_execution_consensus_engine` doesn't sound right because consensus engine also covers post-execution block rewards.

`pre_validate_...` is perhaps not great, but it highlights that returned `true` for a transaction doesn't necessarily mean that the transaction is certainly valid.

Reply:
Or again we must extend conensus_engine methods to include other checks performed outside its scope. Otherwise consensus_engine (as a name which defines a scope) seems cripple to me.

Reply:
> `pre_execution_consensus_engine` doesn't sound right because consensus engine also covers post-execution block rewards.

Right but that is an "action" not a "check"

> 
> `pre_validate_...` is perhaps not great, but it highlights that returned `true` for a transaction doesn't necessarily mean that the transaction is certainly valid.

Ok ... but be my strong objection put on record (LOL)